OpenSSL libraries for .net core support

stefandevo opened the pull request:
In order to use .NET Core, you first need to install the latest version of OpenSSL.

It seemed that this was working without it before, but possible there was some caching going on on some VM's. 

### New Pull Request Checklist

*Please mark the points which you did / accept.*

- [x] The tool I added is stable, and __does not require frequent updates__. _Preinstalled tools on the LTS stacks
  are not updated, so if the tool requires frequent updates you should handle the installation on-demand (see: [Install Any Additional Tool - DevCenter](https://bitrise-io.github.io/devcenter/tips-and-tricks/install-additional-tools/) for more information)_
- [x] I added a version report line to [`system_report.sh`](/system_report.sh) for the new tool(s) I added

viktorbenei commented:
Also worth to mention that OpenSSL is already installed through brew, as a dependency of other brew installed tools. 

Run a brew list on the VMs to get a full list of brew installed packages. 

That said, it'd probably be a good idea to add brew list to the system report too.

viktorbenei commented:
Question: isn't the NET Core available as a brew package? That should solve dependency issues, as brew packages specify the required dependency packages for themselves.

stefandevo commented:
@viktorbenei no, .net core is not available as a brew package. 
Thing is, the current box does not work for .net core. I tested this before (before I made the previous PR to include .net core) but it seemed I was testing this on a box where I added these extra lines before). 
I now tested the new Xamarin stack I saw that .net core was not working as it should be. Adding the lines as explained on [https://www.microsoft.com/net/core#macos](https://www.microsoft.com/net/core#macos) concerning OpenSSL did fix the build error.

Following error is generated on the current Xamarin stack when building .net core projects:

```
Retrying 'FindPackagesByIdAsync' for source 'https://api.nuget.org/v3-flatcontainer/netstandard.library/index.json'.
  The type initializer for 'System.Net.Http.CurlHandler' threw an exception.
    The type initializer for 'Http' threw an exception.
    The type initializer for 'HttpInitializer' threw an exception.
    The type initializer for 'CryptoInitializer' threw an exception.
    Unable to load DLL 'System.Security.Cryptography.Native.OpenSsl': The specified module could not be found.
     (Exception from HRESULT: 0x8007007E)
  Retrying 'FindPackagesByIdAsync' for source 'https://api.nuget.org/v3-flatcontainer/netstandard.library/index.json'.
  The type initializer for 'System.Net.Http.CurlHandler' threw an exception.
    The type initializer for 'Http' threw an exception.
    The type initializer for 'HttpInitializer' threw an exception.
    The type initializer for 'CryptoInitializer' threw an exception.
    Unable to load DLL 'System.Security.Cryptography.Native.OpenSsl': The specified module could not be found.
     (Exception from HRESULT: 0x8007007E)
```

When adding a script step before the build process:

```
brew update
brew install openssl
mkdir -p /usr/local/lib
ln -s /usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib /usr/local/lib/
ln -s /usr/local/opt/openssl/lib/libssl.1.0.0.dylib /usr/local/lib/
```

Made it work.

I saw that the playbook.yml file already had the `brew update` command so that's why I just added the other lines to the xamarin playbook.

So you are right that the /usr/local is probably not correct and should an environment variable. Can you suggest one? $HOME ?

I did a small test with the test project by only adding the two last lines, this seems to work as well, but not sure if the openssl was now cached on this box or not.

I added support to the [https://www.bitrise.io/app/01532ef2773242fe#/builds](https://www.bitrise.io/app/01532ef2773242fe#/builds) script for you to check if you want.

Please advise further how we can make .net core work on bitrise. Thanks.

viktorbenei commented:
> So you are right that the /usr/local is probably not correct and should an environment variable. Can you suggest one? $HOME ?

The issue is not /usr/local. That's fine. The issue is the hardcoded version number (e.g. `libcrypto.1.0.0.dylib`) which will break eventually, and this is something what's managed by brew and package dependencies when a package is installed from brew, that's why `brew` is preferred for pretty much every dependency.

> not sure if the openssl was now cached on this box or not.

The base brew packages are not updated weekly, only when there's an Xcode update. Only the Xamarin specific bits are updated every week(end) -> https://discuss.bitrise.io/t/how-are-the-bitrise-io-stacks-updated-and-with-what-frequency/281

> The Xamarin Stacks are built on top of the latest stable Xcode stack, so the preinstalled things of the base Xcode stack are not updated and are the same as in the base Xcode stack. The Xamarin specific bits are applied on top of the base Xcode image, every week, when generating the Xamarin stacks (using this Ansible playbook).

---

> Please advise further how we can make .net core work on bitrise. Thanks.

To be honest, I'd rather just remove .Net Core from the VM until it gets available as a `brew` package; otherwise it'll break from time to time, which might prevent or delay stack updates. It's not a terrible thing at all, there are better ways to handle fast-changing dependencies than baking it into a semi-frequently updated base platform (keep in mind the preinstalled tools are expected to be stable long term - see the `he tool I added is stable, and does not require frequent updates. Preinstalled tools on the LTS stacks are not updated, so if the tool requires frequent updates you should handle the installation on-demand (see: Install Any Additional Tool - DevCenter for more information)` checklist item in the PR).

The preferred way to handle these types of dependencies is to provide a Step for it, until things settle. That way it's possible to push an update any time, multiple times a day if required (vs the fix update frequency of the VMs).

What do you think? Is there a reason why this should not be transformed into a Step, and should be pre-installed instead? AFAIK the installation is pretty fast, so it shouldn't be an issue, and doing it that way would provide way better flexibility in terms of changing/fixing things.

stefandevo commented:
The hard dependency on these version numbers are coming from the official installation page of .net core for mac [https://www.microsoft.com/net/core#macos](https://www.microsoft.com/net/core#macos).

I added the "use .net core" in the Cake build script so it's optional, but if you don't use Cake and just the regular Xamarin archive step, builds will not run when you include a .net core project type that you want to reference in your app. So this is an issue.

It's possible to create a step "Install .NET Core" with these commands, yes indeed, but I thought it would be more handy for the Xamarin users that it's  "just there" to use, as Mono is just there on the Xamarin boxes.

And of course .net core will evolve over time, but also Xamarin is bringing changes every couple of weeks.

But again, it's your decision :-)

stefandevo commented:
Further thinking of this... I will provide a Step for this.
So you better remove .net core again from the xamarin boxes then :-)

viktorbenei commented:
Yeah, sorry, it was our mistake, we should have suggested you to do it as a step when you sent in the PR for preinstall, I guess we were just too excited 😂

We'll definitely consider preinstalling core once it lands in brew. Until that a step is definitely a better solution, and maybe even after that, if eg the step has an input  for the dotnet core version number 😉

viktorbenei commented:
If you send in the step, we'll highlight it in the weekly changelog as the replacement solution (for the removed preinstalled version)

viktorbenei commented:
And of course thanks for your time & work, and sorry for the inconvenience

stefandevo commented:
https://github.com/bitrise-io/bitrise-steplib/pull/981

viktorbenei commented:
@stefandevo  PR merged - we just changed the ID of the step to `install-dotnetcore` -> should be available in the next ~10 mins ;)

I'll go ahead and remove the preinstalled .Net Core from here, which means next week's (this weekend's) Xamarin stacks won't include it anymore, but we'll add a note about this into the weekly stack change changelog, pointing to the new step.

stefandevo commented:
Ok perfect!

viktorbenei commented:
PR to remove the dotNet Core preinstall + point to the replacement step instead: https://github.com/bitrise-io/osx-box-bootstrap/pull/45
